File: sdk/go/passkey.go
package ggid

import (
	"context"
	"fmt"
)

// PasskeyBeginRegistration initiates WebAuthn/Passkey registration.
// Returns the server challenge (credential creation options).
//
// Usage:
//
//	opts, _ := client.PasskeyBeginRegistration(ctx, token, "my-passkey")
//	// Send opts to browser, call navigator.credentials.create()
func (c *Client) PasskeyBeginRegistration(ctx context.Context, accessToken, deviceName string) (map[string]any, error) {
	var result map[string]any
	err := c.post(ctx, "/api/v1/auth/mfa/enroll", map[string]string{
		"type": "webauthn",
		"name": deviceName,
	}, &result)
	if err != nil {
		return nil, fmt.Errorf("passkey begin registration: %w", err)
	}
	return result, nil
}

// PasskeyFinishRegistration completes WebAuthn/Passkey registration.
// The attestationResponse is the result from navigator.credentials.create().
func (c *Client) PasskeyFinishRegistration(ctx context.Context, accessToken, deviceID, attestationResponse string) error {
	return c.post(ctx, "/api/v1/auth/mfa/verify", map[string]string{
		"device_id": deviceID,
		"code": attestationResponse,
	}, nil)
}

// PasskeyBeginLogin initiates WebAuthn/Passkey authentication.
// Returns the server challenge (credential request options).
func (c *Client) PasskeyBeginLogin(ctx context.Context, username string) (map[string]any, error) {
	var result map[string]any
	err := c.post(ctx, "/api/v1/auth/webauthn/login/begin", map[string]string{
		"username": username,
	}, &result)
	if err != nil {
		return nil, fmt.Errorf("passkey begin login: %w", err)
	}
	return result, nil
}

// PasskeyFinishLogin completes WebAuthn/Passkey authentication.
// The assertionResponse is the result from navigator.credentials.get().
func (c *Client) PasskeyFinishLogin(ctx context.Context, assertionResponse string) (map[string]any, error) {
	var result map[string]any
	err := c.post(ctx, "/api/v1/auth/webauthn/login/finish", map[string]string{
		"assertion": assertionResponse,
	}, &result)
	if err != nil {
		return nil, fmt.Errorf("passkey finish login: %w", err)
	}
	return result, nil
}

File: sdk/rust/src/passkey.rs
//! Passkey/WebAuthn registration and authentication API calls.

use serde_json::Value;
use super::client::GGIDClient;
use super::error::GGIDError;

async fn parse_resp(resp: reqwest::Response) -> Result<Value, GGIDError> {
    let status = resp.status();
    let text = resp.text().await.unwrap_or_default();
    if !status.is_success() {
        return Err(GGIDError::Api { status: status.as_u16(), body: text });
    }
    Ok(serde_json::from_str(&text)?)
}

impl GGIDClient {
    /// Begin WebAuthn/Passkey registration. Returns server challenge options.
    pub async fn begin_passkey_registration(
        &self, token: &str, device_name: &str,
    ) -> Result<Value, GGIDError> {
        let body = serde_json::json!({"type": "webauthn", "name": device_name});
        let resp = self.http
            .post(format!("{}/api/v1/auth/mfa/enroll", self.base_url))
            .header("Authorization", format!("Bearer {}", token))
            .header("X-Tenant-ID", &self.tenant_id)
            .json(&body)
            .send().await?;
        parse_resp(resp).await
    }

    /// Finish WebAuthn/Passkey registration.
    pub async fn finish_passkey_registration(
        &self, token: &str, device_id: &str, attestation: &str,
    ) -> Result<Value, GGIDError> {
        let body = serde_json::json!({"device_id": device_id, "code": attestation});
        let resp = self.http
            .post(format!("{}/api/v1/auth/mfa/verify", self.base_url))
            .header("Authorization", format!("Bearer {}", token))
            .header("X-Tenant-ID", &self.tenant_id)
            .json(&body)
            .send().await?;
        parse_resp(resp).await
    }

    /// Begin WebAuthn/Passkey login.
    pub async fn begin_passkey_login(&self, username: &str) -> Result<Value, GGIDError> {
        let body = serde_json::json!({"username": username});
        let resp = self.http
            .post(format!("{}/api/v1/auth/webauthn/login/begin", self.base_url))
            .header("X-Tenant-ID", &self.tenant_id)
            .json(&body)
            .send().await?;
        parse_resp(resp).await
    }

    /// Finish WebAuthn/Passkey login.
    pub async fn finish_passkey_login(&self, assertion: &str) -> Result<Value, GGIDError> {
        let body = serde_json::json!({"assertion": assertion});
        let resp = self.http
            .post(format!("{}/api/v1/auth/webauthn/login/finish", self.base_url))
            .header("X-Tenant-ID", &self.tenant_id)
            .json(&body)
            .send().await?;
        parse_resp(resp).await
    }
}
